packages/tangle/messagestore.go

```diff
@@ -0,0 +1,254 @@
+
+	// DBSequenceNumber defines the db sequence number.
+	DBSequenceNumber = "seq"
+)
+
+// MessageStore represents the storage of messages.
+type MessageStore struct {
+	messageStorage         *objectstorage.ObjectStorage
+	messageMetadataStorage *objectstorage.ObjectStorage
+	approverStorage        *objectstorage.ObjectStorage
+	missingMessageStorage  *objectstorage.ObjectStorage
+
+	Events   *MessageStoreEvents
+	shutdown chan struct{}
+}
+
+// NewMessageStore creates a new MessageStore.
+func NewMessageStore(store kvstore.KVStore) (result *MessageStore) {
+	osFactory := objectstorage.NewFactory(store, database.PrefixMessageLayer)
+
+	result = &MessageStore{
+		shutdown:               make(chan struct{}),
+		messageStorage:         osFactory.New(PrefixMessage, MessageFromObjectStorage, objectstorage.CacheTime(cacheTime), objectstorage.LeakDetectionEnabled(false)),
+		messageMetadataStorage: osFactory.New(PrefixMessageMetadata, MessageMetadataFromObjectStorage, objectstorage.CacheTime(cacheTime), objectstorage.LeakDetectionEnabled(false)),
+		approverStorage:        osFactory.New(PrefixApprovers, ApproverFromObjectStorage, objectstorage.CacheTime(cacheTime), objectstorage.PartitionKey(MessageIDLength, MessageIDLength), objectstorage.LeakDetectionEnabled(false)),
+		missingMessageStorage:  osFactory.New(PrefixMissingMessage, MissingMessageFromObjectStorage, objectstorage.CacheTime(cacheTime), objectstorage.LeakDetectionEnabled(false)),
+
+		Events: newMessageStoreEvents(),
+	}
+
+	return
+}
+
+// StoreMessage stores a new message to the message store.
+func (m *MessageStore) StoreMessage(msg *Message) {
+	var msgIsNew bool
+	var cachedMsgMetadataObject objectstorage.CachedObject
+	messageID := msg.ID()
+
+	// store message
+	cachedMsg := m.messageStorage.ComputeIfAbsent(msg.ObjectStorageKey(), func(key []byte) objectstorage.StorableObject {
+		msgIsNew = true
+
+		// store the message metadata
+		cachedMsgMetadataObject = m.messageMetadataStorage.Store(NewMessageMetadata(messageID))
+
+		msg.Persist()
+		msg.SetModified()
+		return msg
+	})
+	if !msgIsNew {
+		cachedMsg.Release()
+		return
+	}
+
+	cachedMessage := &CachedMessage{CachedObject: cachedMsg}
+	cachedMsgMetadata := &CachedMessageMetadata{CachedObject: cachedMsgMetadataObject}
+
+	// TODO: approval switch: we probably need to introduce approver types
+	// store approvers
+	msg.ForEachStrongParent(func(parent MessageID) {
+		m.approverStorage.Store(NewApprover(parent, messageID)).Release()
+	})
+
+	// trigger events
+	if m.missingMessageStorage.DeleteIfPresent(messageID[:]) {
+		m.Events.MissingMessageReceived.Trigger(&CachedMessageEvent{
+			Message:         cachedMessage,
+			MessageMetadata: cachedMsgMetadata})
+	}
+
+	// messages are stored, trigger MessageStored event to move on next check
+	m.Events.MessageStored.Trigger(&CachedMessageEvent{
+		Message:         cachedMessage,
+		MessageMetadata: cachedMsgMetadata})
+
+	cachedMessage.Release()
+	cachedMsgMetadata.Release()
```

Comment:
Just for my understanding: Releasing it here potentially means that the message might be in cache for the full data flow?